### ISSUE TITLE
False positive `unused-variable` for homonyms between comprehension variable and outer container

### ISSUE THREAD

**Issue by jacobtylerwalls** [PyCQA/pylint#6136]
### Bug description

```python
def f():
    project_id = []  # this ought to be called project_ids, but...
    assert (project_id > 0 for project_id in project_id)
```

### Command used

```shell
pylint z.py
```

### Pylint output

```shell
************* Module z
z.py:2:4: W0612: Unused variable 'project_id' (unused-variable)

-------------------------------------------------------------------
Your code has been rated at 6.67/10 (previous run: 10.00/10, -3.33)
```

### Expected behavior

No msg

### Pylint version

Regression in #6073. Found it by diffing primer package messages.

**Comment by jacobtylerwalls** [comment]
Not a regression if we fix it before 2.13.5.
